src/main/java/net/openhft/chronicle/network/internal/AddressCache.java

```diff
@@ -0,0 +1,110 @@
+ *   <li><em>30:</em> InetSocketAddresses will be cached for 30s after their first lookup</li>
+ *   <li><em>any negative value:</em> InetSocketAddresses will be cached forever</li>
+ * </ul>
+ * <p>
+ * Note: This will default to caching addresses forever when no value is specified for <em>networkaddress.cache.ttl</em>,
+ * as per the documentation, that default may differ from the JVM default when no security manager is installed.
+ */
+public class AddressCache {
+
+    private static final String NETWORK_ADDRESS_CACHE_TTL = "networkaddress.cache.ttl";
+    private static final int NEVER_EXPIRE = -1;
+
+    private final ConcurrentHashMap<String, CacheEntry> cachedAddresses;
+
+    public AddressCache() {
+        cachedAddresses = new ConcurrentHashMap<>();
+    }
+
+    public InetSocketAddress lookup(String description) {
+        final CacheEntry cacheEntry = cachedAddresses.get(description);
+        return cacheEntry != null ? cacheEntry.address() : null;
+    }
+
+    public void add(String description, String hostname, int port) {
+        cachedAddresses.put(description, new CacheEntry(hostname, port));
+    }
+
+    public void clear() {
+        cachedAddresses.clear();
+    }
+
+    private class CacheEntry {
+        @NotNull
+        private final String hostname;
+        private final int port;
+        private InetSocketAddress inetSocketAddress;
+        private long expiryTime;
+
+        public CacheEntry(@NotNull String hostname, int port) {
+            this.hostname = hostname;
+            this.port = port;
+        }
+
+        /**
+         * Get the {@link InetSocketAddress} for this entry, creating a new one if necessary
+         * <p>
+         * Note: it is important this method stays synchronized, to prevent unnecessary InetSocketAddress creation
+         * but also to ensure any updates to the fields are visible to all threads.
+         *
+         * @return The {@link InetSocketAddress}
+         */
+        public synchronized InetSocketAddress address() {
+            if (isCachingEnabled()) {
+                if (inetSocketAddress == null || (expiryTime >= 0 && System.currentTimeMillis() > expiryTime)) {
+                    inetSocketAddress = createInetSocketAddress(hostname, port);
+                    expiryTime = calculateExpiryTime();
+                }
+                return inetSocketAddress;
+            } else {
+                return createInetSocketAddress(hostname, port);
+            }
+        }
+    }
+
+    /**
+     * Calculate the expiry time for the cache entry using <em>networkaddress.cache.ttl</em>
+     * <p>
+     * See https://docs.oracle.com/javase/7/docs/technotes/guides/net/properties.html for details
+     *
+     * @return The expiry time in milliseconds since the epoch or -1 to never expire
+     */
+    private long calculateExpiryTime() {
```

Comment:
Could not these methods be moved into the CacheEntry and then CacheEntry could be a static class?

Reply:
Yep, they moved around a bit in the process. That is a sensible suggestion.